util/expensivequery/memory_usage_alarm.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expensivequery
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	rpprof "runtime/pprof"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/disk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type memoryUsageAlarm struct {
+	err                    error
+	isServerMemoryQuotaSet bool
+	serverMemoryQuota      uint64
+	lastCheckTime          time.Time
+
+	tmpDir              string
+	lastLogFileName     []string
+	lastProfileFileName [][]string // heap, goroutine
+}
+
+func initMemoryUsageAlarmRecord() (record *memoryUsageAlarm) {
+	record = &memoryUsageAlarm{}
+	if alert := config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio; alert == 0 || alert == 1 {
+		record.err = errors.New("close memory usage alarm recorder")
+		return
+	}
+	if quota := config.GetGlobalConfig().Performance.ServerMemoryQuota; quota != 0 {
+		record.serverMemoryQuota = quota
+		record.isServerMemoryQuotaSet = true
+	} else {
+		// TODO: Get the memory info in container directly.
+		record.serverMemoryQuota, record.err = memory.MemTotal()
```

Comment:
How abouot container environment?

Reply:
Now the only way to use it is to set server-memory-quota config.